# Non-abundant sums
## Problem 23

A perfect number is a number for which the sum of its proper divisors is exactly equal to the number. For example, the sum of the proper divisors of 28 would be 1 + 2 + 4 + 7 + 14 = 28, which means that 28 is a perfect number.

A number n is called deficient if the sum of its proper divisors is less than n and it is called abundant if this sum exceeds n.

As 12 is the smallest abundant number, 1 + 2 + 3 + 4 + 6 = 16, the smallest number that can be written as the sum of two abundant numbers is 24. By mathematical analysis, it can be shown that all integers greater than 28123 can be written as the sum of two abundant numbers. However, this upper limit cannot be reduced any further by analysis even though it is known that the greatest number that cannot be expressed as the sum of two abundant numbers is less than this limit.

Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.

In [4]:
from math import sqrt

def divisor(nb):
    list_of_divisor = [1]
    
    c = sqrt(nb)
    if c.is_integer() and c!=1:
        c = int(c)
        list_of_divisor.append(c)
    else:
        c = int(sqrt(nb))+1
        
    for q in range(2,c):
        if nb%q == 0: #nb = b*q
            list_of_divisor.append(q)
            list_of_divisor.append(nb//q)
    return list_of_divisor


def d(nb):
    return sum(divisor(nb))

def is_abundant(nb):
    if d(nb)>nb:
        return True
    else:
        return False

# Méthode 1

In [5]:
list_of_abundant = [k for k in range(1,28123) if is_abundant(k)]

In [6]:
print(len(list_of_abundant))

6965


Il y a 6965 nbr abondant inférieurs a 28123.

Ainsi on peut donc déterminer l'ensemble des nbr qui peuvent etre ecris comme la somme de deux nbr abondant. Ils sont de la forme _ + _. Avec n choix pour le premier et n pour le 2e ainsi que l'ordre qui ne compte pas par communtativité de la loi + donc il faut retrancher 2. Ainsi il y a $\frac{n^{2}}{2}$

In [7]:
list_of_sum = []

l = len(list_of_abundant)

a = 0
while a<=l:
    
    b=a
    while  b<=l:
        c = list_of_abundant[a] + list_of_abundant[b]
        b+=1
        if c>28123:
            a+=1
            b=a
        elif c not in list_of_sum :
            list_of_sum.append(c)
    a+=1

KeyboardInterrupt: 

In [10]:
len(list_of_abundant)

6965

$$\sum_{k=a+b, (a,b)∈Ab}k = \sum_{k∈ℕ}k - \sum_{k=a+b, (a,b)∉Ab}k = \sum_{k=0}^{28123}k - \sum_{k=a+b, (a,b)∉Ab}^{28123}k $$

$$\sum_{k=a+b, (a,b)∈Ab}k = \frac{n \cdot (n+1)}{2} - \sum_{k=a+b, (a,b)∉Ab}^{28123}$$

In [87]:
solution = ((28123)*(28124))//2 - sum(list_of_sum)

In [88]:
solution

-227202587074

In [101]:
def goto(linenum):
    global line
    line = linenum

line = 1
while True:
    if line == 1:
        response = input("yes or no? ")
        if response == "yes":
            goto(2)
        elif response == "no":
            goto(3)
        else:
            goto(100)
    elif line == 2:
        print("Thank you for the yes!")
        goto(20)
    elif line == 3:
        print("Thank you for the no!")
        goto(20)
    elif line == 20:
        break
    elif line == 100:
        print("You're annoying me - answer the question!")
        goto(1)

yes or no?  


You're annoying me - answer the question!


yes or no?  


You're annoying me - answer the question!


yes or no?  no


Thank you for the no!


In [3]:
is_abundant(12)

True